ocaml/xapi/xapi_vgpu.ml

```diff
@@ -17,9 +17,31 @@ open D
 (* Mutex to prevent duplicate VGPUs being created by accident *)
 let m = Mutex.create ()
 
-(* Only allow device = "0" for now, as we support just a single vGPU per VM *)
-let valid_device device =
-  device = "0"
+(* 1. Since multiple vgpus support, only device number = 0 or specify [11,31] since demu restriction.
+   2. device = 0 will requires xapi to give a valid number
+    - First sort the device number and the trying to find the next valid from range *)
+
+let min_valid_vgpu_device = 11
+let max_valid_vgpu_device = 31
+
+let range low high =
+  let rec aux low high =
+    if low > high then [] else low :: aux (low+1) high in
+  aux low high
+
+let all_valid_devices = range min_valid_vgpu_device max_valid_vgpu_device
+
+let get_valid_device ~__context ~device ~vM =
+  let d = int_of_string device in
+  if d >= min_valid_vgpu_device && d <= max_valid_vgpu_device then device
+  else if d = 0 then begin
+    let all = Db.VM.get_VGPUs ~__context ~self:vM in
+    let all_existing_devices = List.map (fun self -> Db.VGPU.get_device ~__context ~self |> int_of_string) all in
+    try
+      let valid = List.find (fun d -> not (List.mem d all_existing_devices)) all_valid_devices in
+      string_of_int valid
+    with Not_found -> raise (Api_errors.Server_error (Api_errors.invalid_device, [device]))
```

Comment:
This error condition means that we have run out of guest PCI slots, not that device 0 is invalid. So we should introduce a new API error for it.

Reply:
fixed in new commit